### PR TITLE
TM-23 Fail build on compiler warnings

### PR DESCRIPTION
**Pull request by zkiss**
Ability to fail the build on build servers if compiler warnings are introduced.

Includes suppressing the majority of existing warnings. I have only addressed the most trivial issues in order to minimise impact of this PR - I removed unused parameters from private functions in case the change does not leak outside of the class API.

Individual instances might be addressed in followup PRs according to what people find important to address right now.

### CODE REVIEW

File: tools/checkpoint-agent/build.gradle [hunk 1]
```diff
@@ -29,7 +29,7 @@ apply plugin: 'idea'
 description 'A javaagent to allow hooking into Kryo checkpoints'
 
 dependencies {
-    compile "org.jetbrains.kotlin:kotlin-stdlib-jre8:$kotlin_version"
```
Comment (zkiss) [inline tools/checkpoint-agent/build.gradle:32]: Not sure if this has any impact that we care about, but this dependency does cause a warning.

File: core/src/main/kotlin/net/corda/core/contracts/ContractsDSL.kt [hunk 2]
```diff
@@ -42,8 +42,14 @@ fun <C : CommandData> Collection<CommandWithParties<CommandData>>.select(klass:
...
-                filter { if (party == null) true else party in it.signingParties }.
-                map { CommandWithParties(it.signers, it.signingParties, it.value) }
+                filter {
+                    @Suppress("DEPRECATION")
+                    if (party == null) true else party in it.signingParties
+                }.
+                map {
+                    @Suppress("DEPRECATION")
+                    CommandWithParties(it.signers, it.signingParties, it.value)
+                }
```
Comment (zkiss) [inline core/src/main/kotlin/net/corda/core/contracts/ContractsDSL.kt:52]: I think this would be very easy to address but was not 100% sure of logic equivalency, so left it like this.

But I think this should be
```
filter { if (party == null) true else party.publicKey in it.signers }
// and drop the map, because it just makes an identical copy of a value object
```

File: build.gradle [hunk 3]
```diff
@@ -224,7 +230,7 @@ allprojects {
             jvmTarget = "1.8"
             javaParameters = true   // Useful for reflection.
             freeCompilerArgs = ['-Xjvm-default=compatibility']
-            allWarningsAsErrors = project.hasProperty('compilation.allWarningsAsErrors') ? project.property('compilation.allWarningsAsErrors').toBoolean() : false
+            allWarningsAsErrors = project.property('compilation.warningsAsErrors').toBoolean()
```
Comment (zkiss) [inline build.gradle]: rename, so that when I enable it in team city only 4.3+ versions get affected.

File: core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt [hunk 4]
```diff
@@ -107,13 +107,13 @@ class WireTransaction(componentGroups: List<ComponentGroup>, val privacySalt: Pr
...
                     services.networkParametersService.lookup(hashToResolve)
                 },
-                resolveContractAttachment = { services.loadContractAttachment(it) },
                 isAttachmentTrusted = { isAttachmentTrusted(it, services.attachments) }
         )
     }
 
     // Helper for deprecated toLedgerTransaction
     // TODO: revisit once Deterministic JVM code updated
+    @Suppress("UNUSED") // not sure if this field can be removed safely??
```
Comment (zkiss) [inline core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt:117]: would be good to know the answer to this one. question is regarding serialisation backwards compatibility

File: serialization/src/main/kotlin/net/corda/serialization/internal/model/LocalTypeInformationBuilder.kt [hunk 5]
```diff
@@ -51,7 +51,7 @@ internal data class LocalTypeInformationBuilder(val lookup: LocalTypeLookup,
         val previous = validateProperties
         return try {
             validateProperties = false
-            block()
+            this.block()
```
Comment (zkiss) [inline serialization/src/main/kotlin/net/corda/serialization/internal/model/LocalTypeInformationBuilder.kt:54]: this change should be fine, but I don't 100% understand the warning it caused and why this change fixes it.
The warning was 'The expression is unused'

Appears to be a kotlin bug: https://stackoverflow.com/a/54337904

**Review by roastario — DISMISSED** [review]

File: core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt (hunk 4 above)
Comment (zkiss) [reply core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt:117]: @fenryka any ideas on this one?

Comment (mnesbit) [reply core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt:117]: We have to maintain it

**Comment by roastario** [comment]
build

File: core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt (hunk 4 above)
Comment (zkiss) [reply core/src/main/kotlin/net/corda/core/transactions/WireTransaction.kt:117]: Sure. Can I remove the `lazy` block and set the field value to `null`?

**Review by r3domfox — APPROVED** [review]
LGTM
